### PR TITLE
Implement Lightning

### PR DESCRIPTION
**Pull request by Sllambias**
Potential bugs:
- Sampler NEVER used.
- Scheduler warmup num epochs > total num epochs

Big changes:
- [x] Migrate to Lightning
- [x] Migrate to Hydra
- [x] Upgrade Python version to 3.12
- [x] Upgrade package versions
- [x] Remove requirements.txt (deprecated)
- [x] Remove CONSTANTS files. Either set these as defaults in the classes/functions or use the hydra configs to define constants so we only have defaults being set in Configs+Definitions and not Configs+Constants+Definitions.

Dataflow description:
Create Data:
MEDS -> TrainingData, ValData, TestData + FullPopulation

PT:
We take TrainingData + ValData and filter based on subjects in Population. By default FullPopulation but this can be changed for other cohorts.

Create Outcome:
MEDS -> OutcomeA_FullPopulation

FT: 
We take TrainingData + ValData and filter based on subjects in Outcome_for_Population. By default full population for outcome but can be changed for other cohorts. So you can e.g. have OutcomeA_FullPopulation and OutcomeA_OnlyFemales

`create_data` (important) changelog:
 - Aggregation and value creator is removed (should be done in preMEDS, dataset level)  -- only `exclude_regex` is kept
 - `patients_info` is not saved anymore
 - Codding mapping removed (can be made with a custom vocabulary)
 - CLS token removed (never used -- `[CLS]` token is kept in default vocab for prediction tokens)

### CODE REVIEW

File: bonsai/modules/datamodules/base.py
Comment (Sllambias) [inline bonsai/modules/datamodules/base.py]: Remove this?

File: bonsai/configs/core/base_train.yaml [hunk 1]
```diff
@@ -2,7 +2,7 @@ hydra:
   verbose: false
   output_subdir: hydra
   run:
-    dir: ${oc.env:BONSAI_MODELS}/${dataset}/${hydra:job.config_name}/${data.fold}
```
Comment (Sllambias) [inline bonsai/configs/core/base_train.yaml]: Are we sure about removing fold from dir name? How will finetuned models end up not overwriting eachother if doing cv?

File: bonsai/functional/features.py [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+
+    features = df
+    features["age"] = compute_age(features, dob_info)
+    features = exclude_incorrect_event_ages(features)
+
+    features["abspos"] = compute_abspos(features["time"])
+
+    features = features.sort_values(["subject_id", "time"]).reset_index(
+        drop=True
+    )  # TODO: Needed in MEDS?
```
Comment (mikkelfo) [inline bonsai/functional/features.py]: Is `.reset_index(drop=True)` needed? I would think not

File: bonsai/modules/datasets/PretrainDataset.py [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+        masking_ratio: float = 0.8,
+        masking_random_ratio: float = 0.1,
+        masking_ignore_special_tokens: bool = True,
+        cutoff_date: Optional[dict] = None,
+    ):
+        super().__init__(subjects, max_len, background_length, cutoff_date=cutoff_date)
+        self.vocabulary = vocabulary
+
+        self.masking_select_ratio = masking_select_ratio
+        self.masking_ratio = masking_ratio
```
Comment (mikkelfo) [inline bonsai/modules/datasets/PretrainDataset.py]: Might need to explain the difference between `masking_select_ratio` (% of tokens to corrupt) and `masking_ratio` (% of tokens to replace with `[MASK]`) or revise naming

File: bonsai/modules/networks/components/embeddings.py [hunk 4]
```diff
@@ -40,53 +30,54 @@ def __init__(
...
         self.dropout = nn.Dropout(embedding_dropout)
 
         # Initialize embeddings
-        self.concept_embeddings = nn.Embedding(
+        self.code_embedding = nn.Embedding(
             vocab_size, hidden_size, padding_idx=pad_token_id
         )
-        self.segment_embeddings = nn.Embedding(type_vocab_size, hidden_size)
-        self.age_embeddings = Time2Vec(
+        self.segment_embedding = nn.Embedding(type_vocab_size, hidden_size)
```
Comment (mikkelfo) [inline bonsai/modules/networks/components/embeddings.py]: We should actually implement `padding_idx=pad_token_id` and then start `segments` from `1` instead of `0`

File: bonsai/modules/networks/components/embeddings.py
Comment (mikkelfo) [inline bonsai/modules/networks/components/embeddings.py]: Delete this

File: bonsai/modules/create_data.py [hunk 5]
```diff
@@ -0,0 +1,86 @@
+import logging
+from pathlib import Path
+from typing import Optional
+import pandas as pd
+import pyarrow as pa
+from bonsai.functional.features import create_features
+from bonsai.functional.regex_utils import exclude_codes
+
+
+def drop_duplicates(df: pd.DataFrame) -> pd.DataFrame:
```
Comment (mikkelfo) [inline bonsai/modules/create_data.py]: Should we find another place for this in `bonsai.functional`?

File: bonsai/readme.md
Comment (mikkelfo) [inline bonsai/readme.md]: Add a `train.py` readme if we decide that `finetune.py` and `train.py` must be seperated

File: bonsai/run/finetune.py [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+        scheduler_warmup_epochs=cfg.training.scheduler_warmup_epochs,
+        pos_weight=get_loss_weight(
+            cfg.training.loss_weight_function,
+            labels=train_labels,
+        ),
+    )
+
+    ckpt_callback = ModelCheckpoint(
+        dirpath=model_save_dir,
+        monitor="val/loss",
```
Comment (Katja-Jagd) [inline bonsai/run/finetune.py]: Maybe add the monitor parameter into finetune.yaml unless we always use val/loss for fine-tuning ofc.

File: .github/workflows/pipeline.yml
[the remark this answers is not in the document]
Comment (Sllambias) [reply .github/workflows/pipeline.yml]: Yea, we could save a lightweight ckpt and ft from that.

File: bonsai/functional/normalization.py [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+
+def normalize_segments(segments: torch.Tensor) -> List[int]:
+    """Normalize a list of segment IDs to be zero-based and contiguous.
+
+    Takes a list of segment IDs that may have gaps or start from an arbitrary number,
+    and normalizes them to start from 0 and increment by 1 while preserving their relative order.
+
+    Example:
+        >>> normalize_segments([5, 5, 8, 10, 8])
+        [0, 0, 1, 2, 1]
```
[the remark this answers is not in the document]
Comment (Sllambias) [reply bonsai/functional/normalization.py]: So we might not have `[5, 5, 8, 10, 8]` but will still have e.g. `[5, 5, 8, 8, 10]`, right? The current code handles both cases correctly, so I dont see any reason to change anything

[the remark this answers is not in the document]
Comment (mikkelfo) [reply bonsai/functional/normalization.py]: Yeah, I only think the docstring is wrong -- the function works the same
